cpp/include/raft/allocator.hpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_runtime.h>
+#include "cudart_utils.h"
+
+namespace raft {
+
+/**
+ * @brief Interface for an asynchronous device/host allocator.
+ *
+ * An implementation of this interface can make the following assumptions:
+ * - It does not need to be but it can allow async allocate and deallocate.
+ */
+class Allocator {
+ public:
+  /**
+   * @brief Asynchronously allocates a memory region.
+   *
+   * An implementation of this need to return a allocation of n bytes properly
+   * align bytes on the configured device. The allocation can optionally be
+   * asynchronous in the sense that it is only save to use after all work
+   * submitted to the passed in stream prior to the call to allocate has
+   * completed. If the allocation is used before, e.g. in another stream the
+   * behaviour may be undefined.
+   * @todo: Add alignment requirments.
+   *
+   * @param[in] n         number of bytes to allocate
+   * @param[in] stream    stream to issue the possible asynchronous allocation in
+   */
+  virtual void* allocate(std::size_t n, cudaStream_t stream) = 0;
```

Comment:
This is the so called "bare metal" interface. For RAII interface, one should be using `device_buffer` or `host_buffer` classes.

Reply:
I have also explicitly documented this in the doxygen comments for this doc. Let me know what you think.

Reply:
```
template <typename T>
virtual T* allocate(std:size_t n, cudaStream_t stream) = 0;  // allocate sizeof(T) * n
template <typename T>
virtual void deallocate(T* p, std::size_t n, cudaStream_t stream) = 0;
```

Thinking the code above might be easier as users can call
```
auto p = my_allocator.allocate<int32_t>(n, stream);
```
instead of 
```
auto p = reinterpret_cast<int32_t*>(my_allocator.allocate(sizeof(int32_t) * n, stream));
```

Reply:
I don't think we can do templated virtual methods in C++, that easily. If know of the right way to do so, please let me know. Also, I was hoping to keep this interface a bit verbose so as to encourage folks to use the better interface of  `device_buffer` and `host_buffer`.

Reply:
I see, yeah, I didn't know "I don't think we can do templated virtual methods in C++" and I agree with "I was hoping to keep this interface a bit verbose so as to encourage folks to use the better interface of device_buffer and host_buffer"